### ISSUE TITLE
qwen4exp: UD-IQ3_XXS emits incoherent output in every config; dense/iQ controls coherent - defect points to CPU-hosted-expert gather

### ISSUE THREAD

**Issue by qamar-gmail** [spiritbuun/buun-llama-cpp#114]
### Summary

`Qwen3.8-Flash-Next` (UD-IQ3_XXS, 76.3 GiB GGUF from `unsloth/Qwen3.8-Flash-Next-GGUF`) loads
cleanly and streams tokens on commit `7a918624b`, but **every configuration produces incoherent
output** — empty `content` with degenerate repeating `reasoning_content`, or mojibake on the raw
`/completion` endpoint. It reproduces from the **first** request after a clean load.

A dense control (`Qwen2.5-0.5B`, Q4_K_M) is **coherent** through the same binary on both GPU and
CPU, and every low-bit iQ quant the model uses (IQ2_S / IQ3_S / IQ4_NL / IQ3_XS, fork-quantized
from f16 + imatrix) is coherent pure-CPU. All 4 GGUF shards are sha256-identical to HF. The
defect is confined to the qwen4exp MoE decode of this model — most plausibly the top-10-of-512
expert row-gather/multiply from a CPU-hosted (mmap) expert buffer, which only a real multi-GiB
qwen4exp file exercises.

### Environment

- Commit `7a918624b` ("docs: recommend adaptive MoE cache defaults"), CUDA build, sm_120a.
- RTX 5090 (32 GiB) single-GPU (`CUDA_VISIBLE_DEVICES=0`), 62 GiB RAM, i9-12900F, Linux.
- Model `unsloth/Qwen3.8-Flash-Next-GGUF` → `UD-IQ3_XXS` shards 00001–00003 + `MTP/`
  sidecar. 512 experts, 10 routed/token. `per_layer_token_embd` = 26.8 GiB (lazy-read),
  routed experts = 45.3 GiB, dense+attn+output ≈ 4.2 GiB.
- Reason for IQ3_XXS: your README recipe uses **UD-Q4_K_XL**, which does not fit 62 GiB RAM.
  IQ3_XXS is the smallest quant whose experts still fit the host-RAM budget here.

### Repro (minimal, still reproduces)

```bash
CUDA_VISIBLE_DEVICES=0 ./build/bin/llama-server \
  -m Qwen3.8-Flash-Next-UD-IQ3_XXS-00001-of-00003.gguf \
  -ngl 99 -sm layer -fa on -c 32768 -np 1 -ub 512 \
  -ot 'exps=CPU' --fit on --moe-cache off \
  --host 127.0.0.1 --port 8080
```

```bash
curl localhost:8080/v1/chat/completions -H 'Content-Type: application/json' -d '{
  "model":"qwen3.8-flash-next",
  "messages":[{"role":"user","content":"2+2=? Answer with just the number."}],
  "max_tokens":64,"temperature":0}'
```
→ `reasoning_content: "8"`, `content: ""` (wrong answer). Prose prompt →
`reasoning_content: "：\`\`\`\n"`. Output emits **real vocabulary tokens**, so the embedding /
PLE-scoring / output head produce valid logits; the *meaning* collapses downstream of the
routed-expert contribution.

Degenerate signatures seen across configs: empty content + repeating
`travel+booking-booking-booking`, `kawudi/kaqimudi/kaqimudi…`, `yiyiyiyiyi…`,
`lnky1-2-3-4-…-20`, `138.138.138.138`; raw `/completion` returns mojibake
(`├<|i|>┤<|/i|>┤`).

### Config invariance — output stays garbage with each single change

| Variable | Flipped | Result |
|---|---|---|
| MTP draft sidecar (`--spec-type draft-mtp`) | on / **off** | garbage / garbage |
| KV type | `--vbr-entry t8` / default F16 | garbage / garbage |
| Sampling | temperature 0 / 1.0 | garbage / garbage |
| Endpoint | `/v1/chat/completions` / raw `/completion` | garbage / garbage |
| Expert cache | `--moe-cache soft` / **`--moe-cache off`** (pure CPU+mmap experts, no repack) | garbage / garbage |
| PLE lazy-read | `--tensor-read-lazy on` / **`off`** | garbage / garbage (identical `CPU_Mapped` buffers both ways) |

### Controls that pass on the same binary (isolation ladder)

- `Qwen2.5-0.5B-Instruct-Q4_K_M`: coherent, GPU (`-ngl 99`) **and** CPU-under-mmap (`-ngl 0`).
- `Qwen2.5-0.5B` **IQ3_XS / IQ3_S / IQ2_S / IQ4_NL** (all produced by this repo's
  `llama-quantize` from f16 + imatrix): coherent, pure CPU `-ngl 0`.
  ⇒ fork build, server, mmap, GPU & CPU decode, and the exact expert quant kernels are sound.
- All 4 model shards sha256 == HF ⇒ files byte-perfect (download corruption ruled out).
- PLE lazy-read on vs off changes **nothing** (identical mapped buffers, identical output) ⇒
  not a PLE row-read/mapping issue.

### Where this points

The only code path left that no dense control reaches is the **qwen4exp MoE expert gather**:
routing top-10-of-512, gathering expert rows from the 45.3 GiB expert tensor (CPU-hosted under
mmap here — forced, because placing 45 GiB of experts in the 32 GiB VRAM is impossible), then
the multiply / combine. Under the README's tested config those expert tensors live on GPU via
`--moe-cache`; the **CPU-hosted + mmap + low-bit (iq2_s/iq4_nl/iq3_s) expert path may be broken
or unsupported** for qwen4exp — the load log itself warns *"tensor overrides to CPU are used
with mmap enabled — consider using --load-mode none"* (which is not an option at 62 GiB RAM).

### Ask

1. Is CPU-hosted-expert decode (via `-ot 'exps=CPU'` + mmap + low-bit iQ experts) a supported
   path for qwen4exp, or is a GPU expert repack assumed? UD-Q4_K_XL (your validated quant) is
   too large for 62 GiB RAM, so every quant that fits this box forces the CPU-expert path.
2. Is there a smaller qwen4exp / Flash-Next GGUF you can point me at to repro against, or a
   debug flag (`--log-verbose` level, dump logits/expert routing) that would isolate whether
   the expert contribution is even reaching the output?
3. Happy to run any instrumented build you suggest and report back.

Full local writeup with every test and flag source reference: the project's `RUNBOOK.md`.

**Comment by qamar-gmail** [comment]
**Update — this is very likely a quant problem, not a decode bug.**

The author's own buun-llama-cpp benchmark card for the *same* model family uses the **identical**
host-resident-expert config and is coherent + fast:

- Model: `Qwen3.8-Flash-Next-UD-Q6_K_XL` + shared Q8_0 MTP, 2x RTX 4090, 128K ctx
- Config: `--moe-cache soft`, `-ot 'exps=CPU'`, `-sm layer -ngl all`, `-t 16 -tb 16`,
  `-b 512 -ub 128`, `--fit on`, KV q8_0
- Result: coherent long-form run, 38.78 tok/s sustained over 107K tokens, 78% MTP acceptance

`-ot 'exps=CPU'` + `--moe-cache soft` is therefore the **designed, validated path** — that
exonerates the CPU-hosted-expert gather hypothesis from my report.

The delta between that working run and my failing load is the **quant**:

| | Validated (works) | Mine (garbage) |
|---|---|---|
| Quant | **UD-Q6_K_XL** | UD-IQ3_XXS |
| Expert types | high-fidelity K/X | **iq2_s / iq4_nl / iq3_s** |

My dense 0.5B controls prove those exact iq quant kernels decode fine *on a dense model*. But
the failing artifact is specifically unsloth's **UD-IQ3_XXS** GGUF of this qwen4exp model — the
lowest-bit expert quant they publish — which I chose only because Q6_K_XL / Q4_K_XL exceed 62
GiB host RAM here.

**Ask:** has `UD-IQ3_XXS` (or any iq2_s/iq3_s-expert) quant of Qwen3.8-Flash-Next ever been
validated end-to-end on buun-llama-cpp, by unsloth or anyone? If the iq2_s/iq3_s expert quants
themselves are broken (not the decode), that's a silent data-quality issue that would also bite
exllama/vLLM users of the same file — worth flagging upstream to unsloth too. Happy to run any
instrumented build or log-dump you suggest to confirm.

**Comment by spiritbuun** [comment]
I couldn't reproduce this, but, issue 116 points at possible graph behavior differences between CUDA versions - which version are you on?

**Comment by spiritbuun** [comment]
Should be fixed via 3823c9e - give it another try and let me know!
